splice/queries/distribution.py

```diff
@@ -0,0 +1,134 @@
+import hashlib
+import os
+import urllib
+import json
+
+from datetime import datetime
+from collections import defaultdict
+from collections import namedtuple
+
+from sqlalchemy.sql.expression import false
+from splice.models import Tile, Adgroup, Campaign, CampaignCountry
+from splice.web.api.tile_upload import load_bucketer
+
+
+Dists = namedtuple("Dists", ['legacy', 'directory', 'suggested'])
+
+
+def _tile_dict(tile, bucketer, legacy=False):
+    tile_dict = dict(
+        directoryId=tile.id,
+        url=tile.target_url,
+        bgColor=tile.bg_color,
+        title=tile.title,
+        type=tile.type,
+        imageURI=switch_to_cdn_url(tile.image_uri),
+        enhancedImageURI=switch_to_cdn_url(tile.enhanced_image_uri))
+    if not legacy:
+        # only suggested tiles have categories
+        if tile.adgroup.categories:
+            # TODO(najiang@mozilla.com) only grab the first category here.
+            # We should create multiple tiles for multiple categories, with
+            # each tile has a frecent site list based on the category.
+            bucket = bucketer[tile.adgroup.categories[0].category]
+            frequency_caps = {"daily": tile.adgroup.frequency_cap_daily, "total": tile.adgroup.frequency_cap_total}
+            tile_dict.update(dict(
+                titleBgColor=tile.title_bg_color,
+                frecent_sites=bucket["sites"],
+                frequency_caps=frequency_caps,
+                adgroup_name=bucket["adgroup_name"],
+                adgroup_categories=[category.category for category in tile.adgroup.categories],
+                explanation=tile.adgroup.explanation,
+                check_inadjacency=tile.adgroup.check_inadjacency))
+    return tile_dict
+
+
+def switch_to_cdn_url(image_uri):
+    """See https://github.com/oyiptong/splice/issues/203"""
+    from splice.environment import Environment
+
+    env = Environment.instance()
+    basename = os.path.basename(image_uri)
+    return os.path.join(env.config.CLOUDFRONT_BASE_URL, "images/%s" % basename)
+
+
+def get_possible_distributions(today=None, channel_id=None):
+    from splice.environment import Environment
+
+    env = Environment.instance()
+    if today is None:
+        today = datetime.utcnow().date()
+
+    query = (
+        env.db.session
+        .query(Tile)
+        .filter(Tile.paused == false())
+        .filter(Adgroup.paused == false())
+        .filter(Campaign.paused == false())
+        .filter(Campaign.end_date > today)
```

Comment:
I guess end date is not inclusive. If it starts today and ends tomorrow it will only run today. That's fine as long as it's understood everywhere (UI).